lib/db/schema/patch-072-073.sql

```diff
@@ -0,0 +1,239 @@
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+-- Add columns `verified` and `enable` to help manage TOTP token state
+ALTER TABLE `totp`
+ADD COLUMN `verified` TINYINT(1) DEFAULT FALSE,
+ADD COLUMN `enable` TINYINT(1) DEFAULT TRUE,
+ALGORITHM = INPLACE, LOCK = NONE;
+
+-- Add verification details to session token table
+ALTER TABLE `sessionTokens`
+ADD COLUMN `verificationMethod` INT DEFAULT NULL,
+ADD COLUMN `verifiedAt` BIGINT DEFAULT NULL,
+ADD COLUMN `mustVerify` TINYINT(1) DEFAULT TRUE,
+ALGORITHM = INPLACE, LOCK = NONE;
+
+CREATE PROCEDURE `totpToken_2` (
+  IN `uidArg` BINARY(16)
+)
+BEGIN
+
+  SELECT sharedSecret, epoch, verified, enable FROM `totp` WHERE uid = uidArg;
+
+END;
+
+CREATE PROCEDURE `updateTotpToken_1` (
+  IN `uidArg` BINARY(16),
+  IN `verifiedArg` TINYINT(1),
+  IN `enableArg` TINYINT(1)
+)
+BEGIN
+
+    UPDATE `totp` SET verified = verifiedArg, enable = enableArg WHERE uid = uidArg;
+
+END;
+
+CREATE PROCEDURE `verifyTokensWithMethod_1` (
+  IN `tokenIdArg` BINARY(32),
+  IN `verificationMethodArg` INT,
+  IN `verifiedAtArg` BIGINT(1)
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+    -- Update session verification methods
+    UPDATE `sessionTokens` SET verificationMethod = verificationMethodArg, verifiedAt = verifiedAtArg
+    WHERE tokenId = tokenIdArg;
+
+    -- Get the tokenVerificationId and uid for session
+    SET @tokenVerificationId = NULL;
+    SET @uid = NULL;
+    SELECT tokenVerificationId, uid INTO @tokenVerificationId, @uid FROM `unverifiedTokens`
+    WHERE tokenId = tokenIdArg;
+
+    -- Verify tokens with tokenVerificationId
+    CALL verifyToken_3(@tokenVerificationId, @uid);
+
+    SET @updateCount = (SELECT ROW_COUNT());
+  COMMIT;
+
+  SELECT @updateCount;
+END;
+
+CREATE PROCEDURE `sessionWithDevice_12` (
```

Comment:
The procedures below have been updated to return the `mustVerify` property of session tokens from either the session table or unverified tokens table.